src/ore/src/assert.rs

```diff
@@ -14,6 +14,62 @@
 // limitations under the License.
 
 //! Assertion utilities.
+//!
+//! # Optional assertions
+//!
+//! Optional assertions are like debug assertions, but they can be toggled on
+//! and off at runtime in a release build.
+//!
+//! They are useful for guarding conditions that are too expensive to evaluate
+//! in production deployments or load tests, but are worth evaluating in CI.
+//! When disabled, their performance cost is one branch on an atomic.
```

Comment:
I think if we're going to go into this level of detail its worth noting that each branch will be extremely well predicted

Reply:
Two branches. One to check if the lazy_static has been set, one on the actual value.

```
use lazy_static::lazy_static;

lazy_static! {
    pub static ref OPTIONAL_ASSERTIONS: bool = &std::env::var("MZ_OPTIONAL_ASSERTIONS").unwrap() == "asdf";
}

pub fn maybe_panic() {
    if *OPTIONAL_ASSERTIONS {
        panic!("Whoops!");
    }
}
```

Generated code:

```

playground::maybe_panic:
	sub	rsp, 24
	lea	rax, [rip + <playground::OPTIONAL_ASSERTIONS as core::ops::deref::Deref>::deref::__stability::LAZY]
	mov	qword ptr [rsp], rax
	mov	rax, qword ptr [rip + <playground::OPTIONAL_ASSERTIONS as core::ops::deref::Deref>::deref::__stability::LAZY]
	cmp	rax, 3
	jne	.LBB10_1
	mov	rax, qword ptr [rsp]
	cmp	byte ptr [rax + 8], 0
	jne	.LBB10_3
```

Reply:
By the way, this could be dramatically improved by having a global atomic (statically initialized to false), and setting it during main — rather than using the lazy_static mechanism.

But at our stage, such microoptimization would be superfluous.

Reply:
Yeah, I had it like that originally (hence the docstring that claims one atomic read), but the issue is that we want to be able to set this for _every_ test run with `cargo test` and every binary that depends on ore. That means you have to remember to call `ore::assert::init()` in every test and every `main` and empirically that just doesn't happen.

An alternative we could consider is an init function in `ore` using the `ctor` crate, which hacks in support for running functions at boot that aren't explicitly called by `main`: https://github.com/mmastrac/rust-ctor#examples